app/services/account_role_grantor.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+class AccountRoleGrantor
+
+  attr_reader :account, :by
+
+  def initialize(account, by:)
+    @account = account
+    @by = by
+  end
+
+  def grant(user, role)
+    account_user = nil
+
+    within_transaction do
+      # Remove the old owner if we're assigning a new one
+      remove_old_user if role == AccountUser::ACCOUNT_OWNER
+
+      account_user = find_or_build_user_role(user, role)
+      account_user.save!
+
+      # In order to allow changes once an account is closed (i.e. becomes invalid
+      # per the Validator) we cannot run the validations directly on the account.
+      account_user.errors.add(:base, "Must have an account owner") if account.missing_owner?
```

Comment:
I understand that we can’t run validations directly on the account, but can we add this as a validation in `AccountUser`? It seems like the natural place for a custom validation that does this exact check.